src/node/blockstorage.cpp

```diff
@@ -1156,9 +1157,41 @@ FlatFilePos BlockManager::SaveBlockToDisk(const CBlock& block, int nHeight, cons
     return blockPos;
 }
 
+static auto InitBlocksdirXorKey(const BlockManager::Options& opts)
+{
+    // Bytes are serialized without length indicator, so this is also the exact
+    // size of the XOR-key file.
+    std::array<std::byte, 8> xor_key{};
+
+    // Initialize random fresh key, ...
+    FastRandomContext{}.fillrand(xor_key);
+    // ... but allow legacy or user-disabled key.
+    if (!fs::is_empty(opts.blocks_dir) || opts.xor_key == false) std::fill(xor_key.begin(), xor_key.end(), std::byte{});
```

Comment:
Nit: I found this a bit hard to follow and a bit unfortunate that a new random key would be generated every time whether it were used or not. How about:
```diff
@@ -1166,4 +1165,0 @@ static auto InitBlocksdirXorKey(const BlockManager::Options& opts)
-    // Initialize random fresh key, ...
-    FastRandomContext{}.fillrand(xor_key);
-    // ... but allow legacy or user-disabled key.
-    if (!fs::is_empty(opts.blocks_dir) || opts.xor_key == false) std::fill(xor_key.begin(), xor_key.end(), std::byte{});
@@ -1171 +1167,8 @@ static auto InitBlocksdirXorKey(const BlockManager::Options& opts)
-    if (fs::exists(xor_key_path)) {
+    bool key_exists = fs::exists(xor_key_path);
+
+    if (!key_exists && opts.xor_key && fs::is_empty(opts.blocks_dir)) {
+        // Initialize random fresh key, ...
+        FastRandomContext{}.fillrand(xor_key);
+    }
+
+    if (key_exists) {
```

Reply:
Thanks, this is clearer now.